### ISSUE TITLE
Tracking issue for RFC 1940: allow `#[must_use]` on functions

### ISSUE THREAD

**Issue by aturon** [rust-lang/rust#43302]
This is a tracking issue for the RFC "allow `#[must_use]` on functions" (rust-lang/rfcs#1940).

**Steps:**

- [x] Implement the RFC (cc @rust-lang/compiler -- can anyone write up mentoring instructions?)
- [x] Adjust documentation ([see instructions on forge][doc-guide])
- [x] Stabilization PR ([see instructions on forge][stabilization-guide])

[stabilization-guide]: https://forge.rust-lang.org/stabilization-guide.html
[doc-guide]: https://forge.rust-lang.org/stabilization-guide.html#updating-documentation

**Comment by zackmdavis** [comment]
I think I should have an implementation shortly.

**Comment by crumblingstatue** [comment]
I'm just adding a note here that the current implementation of #1940 does _not_ lint for unused `==`, which was the [main motivation](https://github.com/rust-lang/rfcs/pull/1812#issuecomment-265695898) for it.

**Comment by zackmdavis** [comment]
Yeah, that's a little awkward. I can add the comparison operators in a follow-on PR.

**Comment by eddyb** [comment]
Also, this issue shouldn't have been closed. Ahhh I may have not checked for a feature gate in the implementation. I'm a bad reviewer.

**Comment by zackmdavis** [comment]
I have [an implementation](https://github.com/zackmdavis/rust/commit/f222aace0c8bed00dc737286de56044702db7fae) for linting `==` and the other comparison operators. (It's based off the [feature-gating work](https://github.com/rust-lang/rust/pull/43776), so we'll let that land first.)

**Comment by nikomatsakis** [comment]
Hi @zackmdavis / @eddyb -- so I've not had time to check on the PRs, but I wanted to make a request. In preparation for the decision to stabilize, would it be possible for you to collect the various tests into a subdirectory (e.g., `src/test/ui/rfc_1940_must_use_on_functions/`) for easier review? I've found that when deciding whether or not to stabilize, being able to quickly survey how the compiler currently behaves on a bunch of examples is the best way to judge the current state of the feature (and, of course, the number of tests can also give you an idea of how well vetted the implementation is). If tests have already been collected, my apologies!

Either way, can you add a link somewhere (I can copy it into the header of the issue).

Thanks. ❤️

**Comment by nikomatsakis** [comment]
How are people feeling about this? Shall we consider stabilization? Are we using this for the various iterator methods and other purposes that it was envisioned for?

**Comment by kennytm** [comment]
It is currently applied on methods in `PartialOrd` and `PartialEq` only :S

**Comment by nikomatsakis** [comment]
@kennytm was there a plan to use it elsewhere? I forget. If so, is there any reason we are not doing so?

**Comment by kennytm** [comment]
@nikomatsakis There was a plan to add `#[must_use]` to `Result::ok` rust-lang/rfcs#886.

rust-lang/rfcs#1940 only suggests applying to `eq()` and friends (which turns out [isn't enough to fulfill the RFC's motivation](https://github.com/rust-lang/rust/issues/43302#issuecomment-321169882)).  

I don't think there were any plans to use it on iterator methods coming from these two RFCs. IIRC there's some discussion to put `#[must_use]` on `collect()` to avoid newbies using it to consume the whole iterator.

**Comment by Ixrec** [comment]
I definitely wouldn't stabilize this if it's not working for the `==` operator yet, but https://github.com/rust-lang/rust/pull/44103 claims to make it work for `==` and [the playground agrees that it's working](https://play.rust-lang.org/?gist=0a536651a9b67425a31bc6d22ad20462&version=nightly), so the RFC's primary motivation seems fulfilled. I don't think we need to block stabilization on deciding what iterator methods should also have it.

**Comment by scottmcm** [comment]
It looks like linting of `{ x == y; }` is also feature gated right now.  Can we at least stabilize that?

**Comment by oli-obk** [comment]
ping

this seems like it's ready to get stabilized

**Comment by zackmdavis** [comment]
Yes, this feature is very straightforward; the only thing I'd want to consider before stabilization is whether we should add `#[must_use]` to any other standard library methods (like `.collect()`).

(And even then, the only scenarios in which that makes a difference are ones in which we decide to make more functions must-use without stabilizing, and _then_ nightly users point out that we were _wrong_ to do so, but the instability lets us revert without breaking any deny-warnings stable code.)

Give me a few days to find some time to look through the standard library for any other candidate must-use functions, and I'll report back with comments and/or a stabilization PR for 1.25? (Looks like we have until 15 February before the next train leaves.)

**Comment by zackmdavis** [comment]
(Of course, it would also be possible to immediately stabilize _just_ the often-demanded operator lints—which are only under the `fn_must_use` feature gate due to the historical contingency of RFC discussion not sufficiently forseeing implementation details—but continue to evaluate must-use for functions. However, I agree with [the point](https://gist.github.com/mark-i-m/5bccf4dd7d97751eba542fea84b41a9d#stabilize-or-remove) that came up at least a couple times in the Rust2018 discussions that we're sitting on too many unshipped features: ungating all the `fn_must_use` functionality in 1.25 seems reasonable.)

**Comment by zackmdavis** [comment]
There's a perspective that says everything in the standard library that has no side effects that doesn't already return a must-use type (like `Result`) should be must-use. But that feels like a weirdly _excessive_ number of attributes (at least dozens, likely hundreds), right?

**Comment by scottmcm** [comment]
As it's a lint, I'm in favour of just stabilizing and worrying about which functions get it later.

That said, I'd also like to put this on things like clone.  And maybe things like `to_string`, since I saw someone on IRC do `x.to_string();` and wonder why `x` wasn't now a `String`...

**Comment by nikomatsakis** [comment]
I agree we should stabilize. I have no strong opinion about which functions get the lint but I feel like it's ok if it's a lot, if that is useful. =)

@zackmdavis would you be up for preparing a stabilization report? Basically, include a brief summary of what the feature does, and what tests exist for each part of that (so we can verify that it is working correctly). Special attention to edge cases that came up during implementation and how they are being tested or what was decided about them (particularly if it deviates from or was not mentioned in the RFC).

The audience would be people who have been following along, so it doesn't have to explain the feature from first principles -- i.e., no need to delve into motivation per se. (Though a bit more never hurts, e.g. for people coming in from This Week In Rust)

**Comment by zackmdavis** [comment]
(I regret the delay; I'm penciling in some time Saturday to put together a stabilization PR.)

**Comment by Havvy** [comment]
As it's been exactly four months, what's the status on this?

**Comment by zackmdavis** [comment]
The status is that I haven't _forgotten_ that I said I was going to put together a stabilization PR; I'm just terrible at managing my time.  :cry: (And writing [new](https://github.com/rust-lang/rust/pull/48527) [code](https://github.com/rust-lang/rust/pull/48858) is more fun than stabilizing existing code.)

The psychological pressure induced by having been caught breaking my word is probably enough to get me to do it tonight (Pacific time).

**Comment by zackmdavis** [comment]
So, this is unexpected—and _really_ the sort of thing that we would hope to find out _before_ trying to put together the stabilization PR—but we have a serious bug. Building the `getopts` crate with the fn-must-use code (either after ripping out the feature gate, or with the feature attribute inserted) _panics_ (!) when [`qpath_def` gives us](https://github.com/rust-lang/rust/blob/3edb3cc26c4030e62f26fd5f0d9d53fb7b78d618/src/librustc_lint/unused.rs#L80) a `Def::Local` for [this line](https://github.com/rust-lang-nursery/getopts/blob/1733d6d301d4520f3c879086cb40eced498e41d7/src/lib.rs#L1027) and we [try to take its `DefId`](https://github.com/rust-lang/rust/blob/3edb3cc26c4030e62f26fd5f0d9d53fb7b78d618/src/librustc_lint/unused.rs#L91) (which [isn't a real thing](https://github.com/rust-lang/rust/blob/3edb3cc26c4030e62f26fd5f0d9d53fb7b78d618/src/librustc/hir/def.rs#L172-L179)). I haven't found a comprehensibly reduced test case yet.

**Comment by zackmdavis** [comment]
Minimized:

```rust
#![feature(fn_must_use)]

fn main() {
    let c = || { true };
    c();
}
```

yields

```
error: internal compiler error: librustc/hir/def.rs:178: attempted .def_id() on invalid def: Local(NodeId(8))
```

Sadly, this does make me wonder how effective the "let nightly users find bugs in unstable features" strategy is, if so few people have kicked the tires on `#[feature(fn_must_use)]` that no one noticed that it (not even using the functionality, just enabling the flag!) breaks programs containing closures that return a non-unit value.

**Comment by zackmdavis** [comment]
stabilization PR is #48925; separate issue #48926 created for the question of `#[must_use]` elsewhere in the standard library

**Comment by nikomatsakis** [comment]
@zackmdavis do you think you can prepare a "stabilization issue" (or just re-use this issue, if there are no further "pieces" to consider stabilizing). It should look something like this:

https://github.com/rust-lang/rust/issues/48453

That is, summarize briefly what is going on, give a few example tests showing the feature in use in various representative ways (and also highlight any changes that have happened since the RFC). Then I can move to FCP.

**Comment by zackmdavis** [comment]
@nikomatsakis 

It it proposed that we stabilize the code currently under the `fn_must_use` feature gate.

This commits us to the following behaviors:

 * A `#[must_use]` attribute on a function or method will cause the unused-must-use lint to fire when the return value of that function or method is ignored, just (up to the wording of the lint message) like the unused-must-use lint currently (on stable Rust) fires on ignored values whose _type_ has the `#[must_use]` attribute.
     * As for must-use types (the codepath is the same), the attribute can take an optional string literal, which is added to the message. (UI test [source](https://github.com/rust-lang/rust/blob/adf2135adc4a65a78ba053f04c29d7fe0468eb87/src/test/ui/rfc_1940-must_use_on_functions/fn_must_use.rs#L29), [output](https://github.com/rust-lang/rust/blob/adf2135adc4a65a78ba053f04c29d7fe0468eb87/src/test/ui/rfc_1940-must_use_on_functions/fn_must_use.stderr#L19))
     * For trait methods, the attribute must appear on the method signature in the trait _definition_; annotating a trait method _implementation_ is a no-op.
     * Examples of usage or desired usage in the ecosystem:
         * [matrix operations in Servo's geometry library](https://github.com/servo/euclid/blob/94326bc3787fc8645d5e8d6c41e20bb57f1c91e2/src/transform2d.rs#L184)
         * [for C types](https://github.com/briansmith/ring/issues/451)
         * [possibly documenting memory leakage scenarios?](https://github.com/koute/stdweb/pull/127#issuecomment-371630089)

 * Unused values of comparison _operators_ (`<`, `>`, `<=`, `>=` and `==`) are likewise linted.

**Changes from the authorizing RFC 1940**:

 * Many RFC discussants [seemed to expect that](https://github.com/rust-lang/rfcs/pull/1812#issuecomment-265695898) the must-use operator lints would automatically fall out of must-use function lints and marking the comparison trait methods as unused, but the implementation turned out to not work that way. The linting code [handles the operators separately](https://github.com/rust-lang/rust/blob/adf2135adc4a65a78ba053f04c29d7fe0468eb87/src/librustc_lint/unused.rs#L97-L105).

**Status changes since implementation** (#43728, #43776, #44103):

 * The code on master at time of writing actually has a fairly serious bug: if the `fn_must_use` feature is on, programs containing closures that return with non-unit values will ICE (!!).
     * But this is [easy to fix](https://github.com/rust-lang/rust/commit/c4c815d65e601be89e8cb0ae5f1422dd2335d4d2). The patch is included in the stabilization PR #48925, which is blocked on our final-comment-period here.

**Documentation status**:

 * An [unstable book section](https://doc.rust-lang.org/unstable-book/language-features/fn-must-use.html) exists (proposed to be deleted in the stabilization PR, but the text could be cannabalized for other docs).

 * Thanks to @Havvy, there's an [open PR for the Reference](https://github.com/rust-lang-nursery/reference/pull/247) pending our decision here. 

**Unresolved questions and issues** (which need not necessarily block stabilization):

 * Do we want to add `#[must_use]` attributes to functions and methods in the standard libary? (Currently, only the comparison methods `.gt`, `.lt`, _&c._ have the attribute.) If we stabilize now, future `#[must_use]` attributes would be insta-stable. Dedicated issue: #48926
    * Discussed possible uses:    
         * `String.replace` (#48423)
         * `.collect` [(discussion, this issue)](https://github.com/rust-lang/rust/issues/43302#issuecomment-336981160)
         * `Result.ok` and `Result.err` (an explicit motivator of the unapproved [predecessor RFC #886](https://github.com/huonw/rfcs/blob/must-use/text/0000-must-use-functions.md))

 * Should the compiler warn when `#[must_use]` appears on a trait _implementation_, rather than the declaration? (This is currently a no-op, but users might expect it to work and be confused when it doesn't.) ([discussion on #48486](https://github.com/rust-lang/rust/issues/48486#issuecomment-368173610))

 * Maybe the further explanation from the optional string literal that the attribute can take would look better æsthetically and perhaps be more visible in a [note](https://github.com/rust-lang/rust/blob/ca6a98426192f838cc90a18709f92d425b86029e/src/librustc_errors/diagnostic.rs#L167) rather than being appended to main lint message?

```
warning: unused return value of `foo` which must be used: further info
  --> $DIR/fn_must_use.rs:67:5
   |
LL |     foo();
   |     ^^^^^^
```

_vs._

```
warning: unused return value of `foo` which must be used
  --> $DIR/fn_must_use.rs:67:5
   |
LL |     foo();
   |     ^^^^^^
   |
note: further info
```

**Comment by SimonSapin** [comment]
> Should the compiler warn when #[must_use] appears on a trait implementation, rather than the declaration? (This is currently a no-op, but users might expect it to work and be confused when it doesn't.)

Can we make this attribute a compilation error when used in places where it would be a no-op? This leaves the option of making it work later without changing the meaning of programs that compile.

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

I propose that we **stabilize** must-use lints. @zackmdavis has prepared an [excellent summary] here.

However, I am also in favor of @SimonSapin's change to report an error if `#[must_use]` is attached to a method in a trait impl (rather than having that be a no-op), and I'd prefer not to stabilize until we agree to that change as well (though this could be added to the stabilization PR).

[excellent summary]: https://github.com/rust-lang/rust/issues/43302#issuecomment-373967124

**Comment by golddranks** [comment]
Is there any issue that tracks which std APIs are going to have this annotation? I just shot myself into foot with `Write::write()` – I checked the `Result` with `?`, but ignored the amount of the written bytes it returned. I should have used `write_all()`; if it warned me about ignoring the return value, I would have noticed it and it would have saved me from a serious bug.

Another thought: is the annotation going to work with APIs that return `Result`, but where even the happy path value needs not to be ignored? (I.e. `write()` here) I would imagine that it would be quite possible to do this with calls like `write()?`, but the general case with all kinds of combinators before returning the error is much much harder.

**Comment by oli-obk** [comment]
Clippy has a correctness lint for that: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#unused_io_amount

I don't think it can be caught by must_use, but a targeted lint can.

**Comment by zackmdavis** [comment]
> Is there any issue that tracks which std APIs are going to have this annotation?

Please, contribute your thoughts on  #48926.

**Comment by TimNN** [comment]
@nikomatsakis: It looks like it was decided to stabilize this, can I assume that #48925 is now un-blocked?

**Comment by nikomatsakis** [comment]
@TimNN yep

**Comment by SimonSapin** [comment]
> `#[must_use]` on trait-impls (and other non-ADT locations) is already a no-op on stable

This is the situation I was hoping to prevent, I didn’t realize it was already the case.

**Comment by Havvy** [comment]
@aturon The stabilization PR has already gone through. I'm not sure if all the necessary docs are updated yet, although the reference has been.

**Comment by Centril** [comment]
@Havvy is there anything left here to do?

**Comment by Havvy** [comment]
I don't know. It's documented in the reference and there's an issue for the API guidelines book. But I don't think it's talked about in TRPL and definitely untouched in Rust By Example, if it should even be in there.

**Comment by Centril** [comment]
Closing as completed.
